[Beck] Step4 - Add Photo

SangHwi-Back opened the pull request:
# 사진 추가하기 기능 추가(완료: 2022.03.15 19:30)

## 기능 설명

* Add 버튼 옆에 Photo 버튼을 추가하여 사진을 배경이미지로 가지는 뷰를 생성할 수 있게 되었습니다.
* 사진은 사진 앱을 통해 가져오고 메인 뷰가 시작될 때 권한 설정에 대해 물어봅니다.
* 권한 설정이 되어 있지 않을 경우 Photo 버튼은 사용할 수 없습니다.

## 구현 방법

hasViewProperty 라는 프로퍼티를 Model 객체에 추가하였습니다.

Plane의 `addRectangle(with:)`가 전달받게 되면, 파라미터로 전달받은 데이터가 뷰와 관련이 있건 없건 간에 뷰와 관련된 프로퍼티가 있다는 표시를 hasViewProperty를 true로 처리하고 데이터는 뷰 컨트롤러로 NotificationCenter를 통해 전달합니다.

MainScreenViewController는 Plane에서 알림이 오게 되면 hasViewProperty가 참일 경우 이미지 뷰를 추가하여 이미지를 적용하려 합니다. 그렇지 않다면 기존대로 배경색상을 적용합니다.

모든 기능은 동일하지만, 무작위 색상을 설정할 경우도 모델의 hasViewProperty가 참일 경우는 버튼이 아무 작동도 하지 않게 처리하였습니다.

## hasViewProperty는 모델을 제외한 모든 객체에 닫혀 있어야 할까 고민해 보았습니다.

hasViewProperty는 모델이라는 생각으로 작업을 진행하였습니다. 하지만, Xcode 에서 hasViewProperty가 사용된 부분을 검색해보면 실제 사각형들이 나타나는 MainScreenViewController에서만 호출합니다.

결국 뷰 컨트롤러가 한번 쯤 뷰 생성에 관여하는 모델 데이터를 검증하는 과정이 필요하다고 느꼈지만, 지금까지 계속 MVC 구조를 제대로 나누지 못한 것 때문에 고민을 덜 한 것은 아닌지 고민입니다.

SangHwi-Back committed:
Update README Step4

Step4에서 수행해야 할 작업들을 정리해보았습니다.

SangHwi-Back committed:
Refactor for step3 complement tasks

1. Plane 객체의 MainSceneDelegate는 한번만 호출되기 때문에
   MainViewController가 MainScreenViewController를 호출할 때 필요한 작업을 수행하도록 함.
2. 위의 작업을 통해 MainViewController와 Plane의 직접적인 관계는 끊어졌습니다.
3. MainScreenViewController가 참조하는 MainSceneTapDelegate를 구현한 Plane의
   변수명이나 함수명이 그 의도를 정확히 표시하도록 네이밍을 수정합니다.

SangHwi-Back committed:
Edit MainViewController and model

모델 객체의 프로퍼티를 생성하는 코드가 MainViewController에 들어가는 바람에
코드를 재수정하였습니다. Plane에 지정생성자를 추가하여
사각형 추가 시 참조해야 할 뷰의 폭과 높이를 초기화 시 받습니다.

SangHwi-Back committed:
Edit Plane readable

Plane 코드 일부를 수정하여 읽기 편하게 만들었습니다.

SangHwi-Back committed:
Add functionality addImage

Model 구조체 중 hasViewProperty라는 Boolean 값을 이용하여 true일 경우는
이미지뷰에 Notification을 통해 전달된 이미지를 집어 넣는 구조로 구현하였습니다.

SangHwi-Back committed:
Final Edit before PR

PR 이전에 추가 수정할 부분을 찾아 반영하였습니다.

SangHwi-Back commented:
안녕하십니까 JK. PR 요청 후 코멘트를 일부 수정하였습니다.

수정을 진행하였습니다. 검토 후 머지 혹은 추가 수정 사항에 대한 확인 부탁드립니다.

오늘 하루도 고생 많으셨습니다. 감사합니다.

godrm commented on DrawingApp/DrawingApp/Models/FactoryRectangleProperty.swift:
```diff
@@ -26,7 +26,8 @@ final class FactoryRectangleProperty: RectanglePropertyCreator {
             color: randomColor,
-            alpha: randomAlpha
+            alpha: randomAlpha,
+            hasViewProperty: hasViewProperty
```
언급한 hasViewProperty 라는게 어떤 의도인지 좀 더 살펴보고 의견을 남기겠습니다. 
일단 여기에 추가된 것으로 보면 살짝 불안합니다 ㅎㅎ

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -30,38 +30,58 @@ final class Plane: MainSceneTapDelegate {
-            object: self,
-            userInfo: [PostKey.index: index, PostKey.model: model]
+    // MARK: - Initialize Plane & sceneRect
+    private var sceneRect: RectangleRect!
```
이번 미션이 그림을 그리는 앱이다보니 핵심 로직에서도 화면 구성요소를 표현하는 단어가 많아지는 게 당연한 것 같습니다. 모델로 사용하는 `RectangleRect` 도 `사각형사각` 같은 느낌이라 어색합니다. 
이전 PR에서도 타입 이름에 대한 피드백이 많았던 것 같은데, 뷰와 모델이 서로 연관되어 있지만 이름으로 혼란을 주지 않도록 표현하는 것은 중요하니까 다른 사람들이 읽고 혼란스러워하지 않을까 항상 객관적으로 표현해보려고 노력해주세요

SangHwi-Back replied:
넵 알겠습니다. 이 타입을 처음 보는 사람이 있다고 가정하고 다시 생각해보겠습니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -30,38 +30,58 @@ final class Plane: MainSceneTapDelegate {
-            let sceneRect = mainSceneDelegate?.getRect(),
-            let rectangleModel = factory.makeRandomRectangleModel(as: "Subview #\(rectangleModels.count)", rectangleRect: sceneRect)
-        else {
+    func addRectangle(with viewProperty: Any? = nil) {
```
`viewProperty: Any? = nil` 이 부분은 모든 타입이 올 수 있고 심지어 Any? 이라서 nil도 될 수 있네요
들어올 수 있는 값이 엄청 광범위한데 실제로 구체 타입에 추가적인 속성을 넣으려는 것처럼 보이네요. 
사각형을 추가하는 데 이게 어떤 의도인지 아직도 잘 모르겠습니다. 아래를 더 살펴볼게요

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -30,38 +30,58 @@ final class Plane: MainSceneTapDelegate {
-            let rectangleModel = factory.makeRandomRectangleModel(as: "Subview #\(rectangleModels.count)", rectangleRect: sceneRect)
-        else {
+    func addRectangle(with viewProperty: Any? = nil) {
+        guard let rectangleModel = factory.makeRandomRectangleModel(as: "Subview #\(rectangleModels.count)", rect: sceneRect, hasViewProperty: (viewProperty != nil)) else {
```
음.. 생성할 때는 viewProperty가 nil인지 아닌지만 보는군요. 그 값을 쓰는 게 아니라니 조금 더 혼란스럽습니다 ㅎㅎ

SangHwi-Back replied:
단순히 뷰 컨트롤러에서 프로퍼티를 전달하는 상황과 그렇지 않은 상황에서 `addRectangle()`이라는 메소드를 사용할 수 있도록 하려고 한건데 프로퍼티가 너무 말도 안되는 것이 되어버리고 말았습니다.

지금의 상황에서 viewProperty는 PHPickerViewController를 호출하여 생성된 애셋 이미지만을 의미하고 있으니, 의도가 좀 더 명확히 보이는 인터페이스 역할을 하는 메소드를 생성하도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -30,38 +30,58 @@ final class Plane: MainSceneTapDelegate {
         rectangleModels.append(rectangleModel)
         
-        sendNotificationToScreen(using: Plane.addButtonPushed, sends: rectangleModel, at: rectangleModels.endIndex-1)
+        var notificationInfo = [
+            PostKey.index: rectangleModels.endIndex,
+            PostKey.model: rectangleModel
+        ] as [Plane.PostKey : Any]
```
타입 선언과 함께 값을 선언한다면 as로 변환해주기 보다는 `var notificationInfo : [Plane.PostKey : Any] = []` 형태가 더 좋습니다. as는 좀 다른 의미가 됩니다.

SangHwi-Back replied:
as는 하위 타입으로 변환한다는 의미이고 `var notificationInfo: [Plane.PostKey: Any] = []`는 타입 캐스팅이라는 의미이군요.. [AnyHashable: Any]의 의미와는 많이 다른 것 같습니다. 수정하도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -30,38 +30,58 @@ final class Plane: MainSceneTapDelegate {
+        ] as [Plane.PostKey : Any]
+        
+        if let viewProperty = viewProperty {
+            notificationInfo.updateValue(viewProperty, forKey: PostKey.viewProperty)
```
스위프트 사전에 값을 바꿀 때는 `setValue(_, forKey:)`나 updateValue(_, forkey:)` 보다는 subscript를 권장합니다.
여기서 사용한 메소드는 objective-c 기반의 KeyValue coding 이라는 건데 동작은 하지만 권장하지 않는 방식입니다. 
결과적으로는 상위에서 넘겨주는 viewProperty를 noti에 담아서 보내려는 의도인가 보네요. 이러면 또 이걸 받는 부분도 살펴봐야 할 것 같습니다. 
여기까지만 봐도 viewProperty 단지 한 곳이 아니라 여러 곳에 영항을 주는 것 같은데 필요한 값을 context 처럼 전달해주려는 의도라면 괜찮을 수도 있는데, 꼭 필요한지는 옵저버 코드도 살펴봐야 할 것 같네요. 
마치 입력에서 출력까지 어떤 값을 전달하는 모양이네요

SangHwi-Back replied:
viewProperty가 `MainViewController` -> `Plane` -> `MainScreenViewController` 로 전달되고 있습니다.

`MainViewController` -> `Plane`, `Plane` -> `MainScreenViewController` 두 개를 이용해서 구조 내 요소들이 가지는 결합을 더 느슨하게 만들어보겠습니다.
넓게 보지 못하는 습관이 또 이상한 코드를 만들었던 것 같습니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -71,32 +91,28 @@ final class Plane: MainSceneTapDelegate {
         
-        var planeNoti = Notification(name: Plane.rectangleViewTouched, object: self, userInfo: nil)
+        var noti = Notification(name: Plane.selectStatusChanged, object: self, userInfo: [PostKey.index: (selectedIndex ?? -1)])
+        NotificationCenter.default.post(noti) // Plane.selectStatusChanged
+        
+        noti.name = Plane.rectangleViewTouched
         if let index = index, rectangleModels.count-1 >= index {
-            planeNoti.userInfo = [
-                PostKey.index: index,
-                PostKey.model: rectangleModels[index]
-            ]
+            noti.userInfo?.updateValue(rectangleModels[index], forKey: PostKey.model)
+        } else {
+            noti.userInfo = nil
```
didSelect()는 이걸 호출하는 메소드가 아니라 주석처럼 델리게이트에서 호출해주는 콜백인거죠?
그럼 매개변수 index는 상위에서 전달해주는 데 plane 내부의 index와 같다는 보장이 얼마나 있을까요? -1이나 엄청 큰 값이 들어올 수도 있을 것 같네요

SangHwi-Back replied:
넵 알겠습니다. 애플리케이션 구조 내 각 요소들 중 validation할 수 있는 요소들이 있는지 더 찾아보도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/Models/RectangleProperty.swift:
```diff
@@ -19,22 +19,16 @@ final class RectangleProperty: RectanglePropertyCreator {
         self.size = RectSize(width: properties.width, height: properties.height)
         self.rgbValue = color
         self.alpha = alpha
+        self.hasViewProperty = hasViewProperty
```
실제로 값을 가지지 않고 Bool 타입만 있는 게 어떤 의미가 있는지 모르겠습니다.

SangHwi-Back replied:
hasViewProperty 를 제외하고도 과제의 기능을 구현할 수 있을 것 같습니다.

다시 시도해보고 PR 드릴 수 있도록 하겠습니다!

godrm commented on DrawingApp/DrawingApp/Views/MainScreenViewController.swift:
```diff
@@ -101,7 +96,7 @@ final class MainScreenViewController: UIViewController, MainSceneDelegate, UIGes
-        let rect = Rectangle(model: model, index: index)
+        let rect = Rectangle(model: model, index: index, backgroundImage: (userInfo[.viewProperty] as? UIImage))
         self.rectangleViews.append(rect)
         view.addSubview(rect)
```
아.. 사용하는 곳을 보니 알겠네요. Image를 입력 흐름에 넣어서 출력 흐름에 꺼내서 뷰를 생성하네요.
이걸 보니 입력 / 출력 흐름이 분리되어 있지 않은 것고 강하게 결합된 것 같습니다
그리고 데이터 추상화 관점에서도 이미지 데이터를 모델에서 보관하지 않기 때문에 핵심 로직이 이미지를 갖고 있지 않습니다. 
핵심 로직이 있는 모델만 놓고서도 동작이 완벽하게 처리가 되야 합니다. 뷰와 분리해서 모델만으로 테스트 가능하고, 저장도 되야하는거죠. 지금 흐름 상으로는 모델만으로는 Photo 라는 타입을 다룰 방법이 없습니다.
하위 모델이 RectangleProperty 하나 뿐인건가요? 역할 마다 작은 다른 타입을 만드는 건 이상한게 아닙니다. 따로 선언해보면 어떨까요?

SangHwi-Back replied:
로직을 모델에 분리한다.....

이번 수정 및 다음 스텝에서도 유의하도록 하겠습니다.

godrm reviewed (changes requested):
본문에서 우려했던 내용을 살펴봤습니다. 
사진 처리를 위해서 기존 로직에 꽤 많은 영향이 있었네요.
다음 내용을 살펴보시고 개선해봅시다. 다른 타입을 만드는 것부터 시작해보면 어떨까 합니다

SangHwi-Back committed:
Edit image property type handling

메인 뷰 컨트롤러에서 스크린까지 전달되는 이미지 데이터를 Any로 관리하던 것을,
모델 이미지 데이터 구조체를 저장한 상태로 스크린이 가져다 쓰는 방식으로 변경하였습니다.
이를 통해 좀 더 뷰 컨트롤러와 모델 간의 결합력을 느슨하게 하였습니다.

SangHwi-Back committed:
Update Plane interface's validation

Plane객체에서 모델을 관리하는 배열의 인덱스 범위를 벗어나거나,
인터페이스의 의도를 벗어나는 값이 들어올 경우를 추가로 관리할 수 있는 코드를 추가하였습니다.

SangHwi-Back committed:
Rename RectangleRect type in factory

팩토리 객체에서 사용하는 RectangleRect는 중복되는 의미의 단어가 겹쳐져 있으므로,
'스크린의 사각형'이라는 의미인 ScreenSceneRect로 명명합니다.
ViewRect라는 의미도 너무 광범위한 의미라고 생각되어 ScreenRect로 변경하였습니다.

SangHwi-Back committed:
Change location of addProperty

Plane의 addProperty()가 부적절한 위치, 부적절한 접근 제한자를 가지고 있어서
위치를 옮기고 접근 제한자 private를 부여하였습니다.

SangHwi-Back commented:
안녕하세요 JK. 말씀해주신대로 `let viewProperty: Any?`로 처리되었던 부분을 Data 객체로 처리될 수 있게 타입을 변경하였습니다. MainViewController에서 addRectangle()에 다른 데이터를 전달해서 뷰를 만들고 싶을 경우, 그 데이터에 따라 인터페이스를 만들어주면 되도록 함수들을 분리해보았습니다.

이를 통해 `var hasViewProperty: Bool`가 사라지면서 애플리케이션 요소 간 결합력이 느슨해지는 효과가 있었습니다.

그 외에 네이밍 등 지적해주신 부분을 수정했사오니 확인을 부탁드리겠습니다. 감사합니다.

godrm commented on DrawingApp/DrawingApp/Models/FactoryRectangleProperty.swift:
```diff
@@ -27,20 +27,20 @@ final class FactoryRectangleProperty: RectanglePropertyCreator {
             alpha: randomAlpha,
-            hasViewProperty: hasViewProperty
+            backgroundImageData: data
         )
```
생성자가 통합되어서 추상 데이터 타입처럼 동작하는 것 같습니다
이 부분에 대한 피드백은 단지 hasViewProperty 사용하느냐가 아니라, 
hasViewProperty를 넘기는 걸로 Rectangle와 Photo를 둘 다 하나의 데이터 구조로 처리하고 있다는 부분입니다.
포함하는 하위 속성이 다르면 서로 다른 타입으로 분리해서 단일 책임을 지키고, 다형성으로 각각의 동작을 하도록 나눠보라는 겁니다.
기존에 Rectangle에 대한 속성을 다루는 타입이 있으면 그것은 그대로 두고, 새로운 타입을 추가하는 거죠. 그래야 변경에 닫혀있고 확장에 열려있는거니까요. 기존꺼를 자꾸 바꿀꺼면 기존 코드를 사용하는 부분까지 영향을 주게되고 거의 모든 곳을 수정해야 할테니까요.

SangHwi-Back replied:
사각형은 `배경색이 있는 사각형`, `배경 사진이 있는 사각형`의 타입으로 나누고 이들은 공통 슈퍼 클래스를 상속받으면서
슈퍼 클래스는 `인덱싱이 가능한 객체`라는 의미를 가진 프로토콜로 구현해보도록 하겠습니다.

각 사각형 클래스는 SRP를 따르고 모델 역할인 Plane은 관리하는 모델 데이터 배열이 LSP에 의해 한 곳에서 관리될 수 있도록 해 보도록 하겠습니다.

시간이 좀 걸릴수도 있겠지만 잘 구현해서 다시 확인 요청드리도록 하겠습니다. 감사합니다.

godrm commented:
어떤 타입 A를 만들고, 비슷한 타입 B를 만들어야 하는데 A에 속성을 추가해서 이럴때는 A처럼, 저럴때는 B처럼 쓰기 보다는
공통적인 부분을 찾아내서 C를 수퍼 클래스 또는 프로토콜로 선언하고, A와 B를 C에서 확장해서 만들 수 있는 것을 시도해보세요
지금은 기존 타입을 바꾸고 있기 때문에 객체지향이 아니라 데이터 중심 프로그래밍을 하고 있는 셈입니다. 
별개 타입을 만들고 공통 부분을 추상화하고 다른 부분은 다형성으로 동작하도록 시도해보세요. 그건 꼭 해보셔야 할 것 같아요!

SangHwi-Back committed:
Rectangle view and model type expansion

색깔이 있는 사각형, 이미지가 있는 사각형에 대해 모델과 뷰 타입들을 확장하였습니다.
뷰 타입의 경우 사각형의 크기와 위치를 가지는 Rectangle 클래스를 필수로 구현하도록 하였고
모델 타입의 경우 이미지 혹은 색깔을 프로퍼티로 갖고 RectangleProperty 클래스를 구현합니다.
이제부터 새로운 뷰가 추가될 경우
뷰 클래스는 Rectangle, 모델 클래스는 RectangleProperty를 공통적으로 구현해야 합니다.

SangHwi-Back committed:
Edit addRectangleObserver more readable

switch문을 사용하여 코드를 좀 더 읽기 쉽도록 하였습니다.

SangHwi-Back committed:
Update step4 complements wrap time

스텝4의 추가 수정사항 시간을 기입하였습니다.

SangHwi-Back commented:
타입들의 공통 부분을 찾아내서 프로토콜, 슈퍼 클래스는 따로 빼서 서로의 역할을 더 강화하면서 객체지향적으로 구현해보려 노력해보았습니다.

사각형 뷰의 모델인 `RectangleProperty` 클래스는 이제 모든 모델을 대표하고 하위 타입들의 슈퍼 클래스로만 역할합니다.
`RectangleProperty`의 하위 클래스로는 `ColoredRectangleProperty`, `ImageRectangleProperty`가 있습니다.

사각형 뷰를 뜻하는 `Rectangle` 클래스는 위의 모델처럼 모든 사각형 뷰를 대표하고 슈퍼 클래스 역할만 합니다.
`Rectangle`도 위와 비슷한 네이밍으로 `ColoredRectangle`, `ImageRectangle` 하위 클래스가 있습니다.

배경색이 있는 경우는 Colored~, 배경이미지가 있는 경우는 Image~ 접두어를 갖습니다.

추가 수정을 완료하였습니다! 확인 요청드리겠습니다. 감사합니다!

godrm commented on DrawingApp/DrawingApp/Views/MainScreenViewController.swift:
```diff
@@ -96,7 +96,18 @@ final class MainScreenViewController: UIViewController, UIGestureRecognizerDeleg
         }
         
-        let rect = Rectangle(model: model, index: index)
+        var rect: Rectangle!
+        
+        switch model {
+        case let model as ImageRectangleProperty:
+            rect = ImageRectangle(model: model, index: index)
+        case let model as ColoredRectangleProperty:
+            rect = ColoredRectangle(model: model, index: index)
+        default:
+            LoggerUtil.debugLog(message: "Initialize rectangle failed. \(model.name)")
+            return
+        }
```
결국 타입을 생성하는 부분에서는 구체 타입을 알아야 하는데요
그게 뷰 컨트롤러야만 하는 것인가는 고민스러운 부분입니다.
그래서 저는 구체 타입을 알고 생성하는 팩토리가 있는 게 장점이라고 생각하거든요
뷰를 생성할 때도 이렇게 switch-case를 두고 구분하기 보다는 타입이 늘어나도 코드 변경 없이 뷰를 생성할 수 있으면 가장 좋습니다. 한 번 고민해보세요

SangHwi-Back replied:
아하... 맨 마지막에 이런 흠을.... 다음 머지때 수정사항을 추가하도록 하겠습니다. 감사합니다.

godrm commented:
> 타입들의 공통 부분을 찾아내서 프로토콜, 슈퍼 클래스는 따로 빼서 서로의 역할을 더 강화하면서 객체지향적으로 구현해보려 노력해보았습니다.

수고하셨습니다.
